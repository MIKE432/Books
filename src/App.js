import React from 'react';
import './App.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import './ui/styles/StylesToExport.scss'
import RouterComponent from './core/routes/router'
function App() {
  return (
      <RouterComponent/>
  );
}

export default App;
