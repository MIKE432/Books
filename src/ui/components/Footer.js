import React from 'react'

const Footer = () => (
    <div className="footer">
        <h1>Micha≈Ç Raszczuk 2020</h1>
    </div>
)

export default Footer