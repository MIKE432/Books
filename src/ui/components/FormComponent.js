import React from 'react'
import { languages } from '../../core/defaults'
import _ from 'lodash'
import { Text } from '../ctrls/Text'
import { Col, Row } from 'react-bootstrap'

export class FormComponent extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            q: ""
        }
    }
    handleSumit = e => {
        e.preventDefault()

        this.setState({ q: this.state.q, langRestrict: this.state.langRestrict }) 
        this.props.handleSubmit(this.state)
    }

    render() {
        return ( 
            <form className="main-form" onSubmit={this.handleSumit}>
                    <Row>
                        
                        <Col>
                            <Text 
                                value={this.state.q} 
                                onChange={(e) => { 
                                    this.setState({ q: e.target.value })
                                    const filters = { ...this.state, q: e.target.value }
                                    this.props.handleSubmit(filters)
                                }}
                                className="form-control input-error" 
                                name='q' 
                                placeholder="Tytuł"
                                isErrorConditionBlock={value => false} 
                                errorMsg="Brak tytułu"
                            />
                        </Col>
                        <Col>
                            <select 
                                className='browser-default custom-select'
                                name='langRestrict'
                                onChange={(e) => { 
                                    this.setState({ q: this.state.q, langRestrict: e.target.value }) 
                                    const filters = { ...this.state, langRestrict: e.target.value }
                                    this.props.handleSubmit(filters)
                                }}
                                placeholder="Język">
                                {
                                    _.keys(languages).map((language, key) => <option value={languages[language]} key={key} label={language}/>)
                                }
                            </select>
                        </Col>
                        <Col>
                            <select 
                                className='browser-default custom-select'
                                name='langRestrict'
                                onChange={(e) => {
                                    this.setState({ ...this.state, filter: e.target.value }) 
                                    const filters = { ...this.state, filter: e.target.value }
                                    this.props.handleSubmit(filters)
                                }}
                                placeholder="Język">
                                    <option value={null} label='Typ e-booka' />
                                    <option value='free-ebooks' label='Darmowe' />
                                    <option value='paid-ebooks' label='Płatne' />
                                    <option value='ebooks' label='Darmowe i płatne' />
                            </select>
                        </Col>
                    </Row>
            </form>
        )
    }
}