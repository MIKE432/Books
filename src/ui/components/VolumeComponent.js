import React from 'react'
import LinesEllipsis from 'react-lines-ellipsis'

export default (props) => {
    const { 
        title,  
        authors, 
        smallThumbnail, 
        description, 
        publishedDate,
        price,
        currency,
        isForSale,
        isFree,
        Ref } = props

    return (
    <div className="volume-container" ref={Ref}>
        <div className="img-container">
            { smallThumbnail ? <img src={smallThumbnail} alt=""/> : <img src="book.png" alt=""/>}
        </div>
        <div className="additional-info-container">
            <h2><LinesEllipsis text={title} maxLine='2' ellipsis='...' trimRight basedOn='letters'/></h2>
            <h4 className="h4-description">
                <LinesEllipsis text={ description ? description : "Brak Opisu"} maxLine='2'   ellipsis='...' trimRight basedOn='letters'/>
            </h4>
            <h4 id="autors">
                {
                    authors && authors.length === 1 ? "Autor: " : "Autorzy: "
                }
                {
                    authors && authors.map(author => author)
                }
            </h4>
            <h4>
                {
                    publishedDate && "Data wydania: " + publishedDate
                }
            </h4>
            <div className="price">
                {
                    isForSale ? isFree ? <h2 id="free">Darmowe</h2> : <h2 id="for-sale" >
                    Cena: {price} {currency}
                </h2> :
                    <p id="not-for-sale" >Nie na sprzeda≈º</p>
                }
            </div>

        </div>  
    </div>)
}