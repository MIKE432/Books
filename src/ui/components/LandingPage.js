import React from 'react';
// import { restCall } from '../../core/services/restClient'
// import { baseUrl } from '../../core/defaults';
// import VolumeComponent from './VolumeComponent';
import _ from 'lodash';
import Footer from './Footer';

class LandingPage extends React.Component {
    constructor(props) {
        super(props)
        this.titleRef = React.createRef()
        this.headerRef = React.createRef()
        this.bookRef = React.createRef()
        this.rootRef = React.createRef()
        this.state = {
            readingPros: [
                {
                    id: 1,
                    value: "1. Redukuje stres"
                }, {
                    id: 2,
                    value: "2. Poprawia pamięć"
                }, {
                    id: 3,
                    value: "3. Inspiruje"
                }, {
                    id: 4,
                    value: "4. Rozwija wyobraźnię"
                }, {
                    id: 5,
                    value: "5. Poprawia skupienie i koncentrację"
                }, {
                    id: 6,
                    value: "6. Poszerza słownictwo"
                }, {
                    id: 7,
                    value: "7. Zwiększa atrakcyjność dla płci przeciwnej"
                }, {
                    id: 8,
                    value: "8. Stymuluje mózg"
                }, {
                    id: 9,
                    value: "9. Poszerza światopogląd"
                }, {
                    id: 10,
                    value: "10. Rozwija wrażliwość na innych ludzi"
                }, {
                    id: 11,
                    value: "11. Poszerza wiedzę"
                }, {
                    id: 12,
                    value: "12. Zwiększa zdolności analityczne"
                }, {
                    id: 13,
                    value: "13. Poprawia pisanie"
                }, {
                    id: 14,
                    value: "14. Pomaga zasnąć"
                }, {
                    id: 15,
                    value: "15. Kształtuje osobowość"
                }               
            ]
        }
        this.refsForAllElements = this.state.readingPros.reduce((acc, value) => {
            acc[value.id] = {
                id: value.id,
                ref: React.createRef(),
                ratio: 0
            }
            return acc
        }, {})

        const callback = entries => {
            entries.forEach(entry => {
                this.refsForAllElements[entry.target.id].ratio = entry.intersectionRatio
                if(this.refsForAllElements[entry.target.id].ratio > 0)
                    this.refsForAllElements[entry.target.id].ref.current.classList.add("pro-visible")
            })

        }

        this.observer = new IntersectionObserver(callback, {
            root: this.rootRef.current
        })
        
    }

    componentDidMount() {
        _.values(this.refsForAllElements).forEach(value => this.observer.observe(value.ref.current))
    }

    handleClick = () => {
        this.props.history.push("/search")
    }


    render() {
        return (
            <div ref={this.rootRef}>
                <div className="main-header" ref={this.headerRef}>
                    <h1 ref={this.titleRef}>My Liblary</h1>
                </div>
                <div className="landing-theme">

                    <button className="btn btn-success"onClick={this.handleClick}>Wyszukaj ksiąkę!</button>
                </div>
                <div className="reading-information">
                    <i className="fas fa-book fa-7x"></i>
                    <h1>Czytanie:</h1>
                    {
                        this.state.readingPros.map(pro => <p className="pro-invisible"key={pro.id} id={pro.id} ref={this.refsForAllElements[pro.id].ref}>{pro.value}</p>)
                    }
                </div>
                <Footer />
            </div>
        )
    }
}
export default LandingPage;