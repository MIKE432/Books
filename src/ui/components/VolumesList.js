import React, { useState, useEffect, useCallback, useRef } from 'react';
import { baseUrl, elementsPerPage, mapBookToViewModel } from '../../core/defaults';
import VolumeComponent from './VolumeComponent';
import { FormComponent } from './FormComponent'
import axios, { CancelToken } from 'axios'
import LoadingComponent from '../ctrls/LoadingComponent'

const useSearch = (filters, page) => {

    const [books, setBooks] = useState([])
    const [inProgress, setInProgress] = useState(false)
    const [hasMore, setHasMore] = useState(false)
    const [isError, setIsError] = useState(false)
    useEffect(() => {
        setBooks([])
        setIsError(false)
    }, [filters])

    useEffect(() => {

        console.log(filters)
        setInProgress(true)
        setIsError(false)
        if(filters.q === '') return
        let cancel

        axios({
            method: 'GET',
            url: baseUrl,
            params: { q: filters.q, startIndex: elementsPerPage * page, langRestrict: filters.langRestrict, maxResults: elementsPerPage },
            cancelToken: new CancelToken(ct => cancel = ct)

        }).then(result => {
            setInProgress(false)

            if(!result.data.items) {
                setIsError(true)
                setHasMore(false)
                return
            }
            setBooks(prevBooks => [...prevBooks, ...result.data.items.map(mapBookToViewModel)])
            setHasMore(result.data.items.length === elementsPerPage)
            
        }).catch( error => {
            
            if(axios.isCancel(error)) {
                
                return
            }
        })

        return () => cancel()

    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [filters, page])

    return { books, inProgress, hasMore, isError }
}

const VolumesList = (props) => {
    
    const [page, setPage] = useState(1)
    const [filters, setFilters] = useState({ maxResults: elementsPerPage * page, q: '', langRestrict:'' })

    const { books, inProgress, hasMore, isError } = useSearch(filters, page)

    const observer = useRef()
    
    const lastElemRef = useCallback(elem => {
        if(inProgress) 
            return

        if(observer.current) 
            observer.current.disconnect()
        
        observer.current = new IntersectionObserver(entries => {
            if(entries[0].isIntersecting && hasMore)
                setPage(prevPage => prevPage + 1)
        })
        if(elem) observer.current.observe(elem)

    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [inProgress, hasMore])

    const filterChanged = (values) => {
        const q = values.q + ( values.authors !== "" ? "+inauthors=" + values.authors : "")
        setFilters(prevFilters => ({ q: q , langRestrict: values.langRestrict, maxResults: elementsPerPage }))
        setPage(0)
    }

    return (
        <div>
            <FormComponent {...props} handleSubmit={filterChanged}  />

            {
                filters.q !== '' ? <div className="list-container">
                {
                    books ? books.map((book, index) => {
                        if(books.length === index + 1)
                            return <VolumeComponent key={index} Ref={lastElemRef} {...book} />
                        else
                            return <VolumeComponent key={index} {...book} />
                    }) : <span>Brak ksiąek do wyświetlenia</span>
                }
                </div> : 
                <div className="empty-state-component-volumes-list">
                    <p>Wpisz tytuł, aby wyświetlić pozycje</p>
                </div>

            }
            {
                isError ? <div className="empty-state-component-volumes-list">
                            <p>Przepraszamy, nie ma takiej ksiąki na naszej półce</p>
                        </div> : null
            }

            <LoadingComponent isVisible={inProgress && filters.q !== ''} />
            {
                !hasMore && filters.q !== '' && !inProgress && books.length !== 0 ? <div className='center-content'><p>Brak kolejnych wyników</p></div> : null
            }
        </div>
    )
}
export default VolumesList;