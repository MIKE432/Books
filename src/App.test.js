import React from 'react'
import App from './App'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import RouterComponent from './core/routes/router'
import { FormComponent } from './ui/components/FormComponent'
import VolumeComponent from './ui/components/VolumeComponent'

Enzyme.configure({ adapter: new Adapter() })

const appWrapper = shallow(<App />)

it('should render <App />', () => {
  expect(appWrapper)
})


it('app renders RouterComponent', () => {
  expect(
    appWrapper.find(RouterComponent).length,
  ).toBe(1);
});

describe('<FormComponent />', () => {
  it('should render', () => {
    const wrapper = shallow(<FormComponent/>)
    expect(wrapper)
  })

  it('should match snapshot', () => {
    const wrapper = shallow(<FormComponent/>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('First child of FormComponent is div ', () => {
    const wrapper = shallow(<FormComponent />)
  
    expect(wrapper.first().type()).toBe('div')
  })
  
})

describe('<VolumeComponent />', () => {

  it('should render', () => {
    const wrapper = shallow(<VolumeComponent/>)
    expect(wrapper)
  })

  it('should match snapshot', () => {
    const wrapper = shallow(<VolumeComponent/>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('First child of VolumeComponent is div ', () => {
    const wrapper = shallow(<VolumeComponent  />)
  
    expect(wrapper.first().type()).toBe('div')
  })   
})