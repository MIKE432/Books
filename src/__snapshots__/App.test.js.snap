// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormComponent /> should match snapshot 1`] = `
<div
  className="form-main-container"
>
  <div
    className="search-title-container"
  >
    <div
      className="search-title-icon"
      onClick={[Function]}
    >
      <i
        className="fas fa-arrow-left fa-2x hover-icon"
      />
    </div>
    <div
      className="search-title"
    >
      <h1>
        My Liblary
      </h1>
      <h3>
        Wyszukaj
      </h3>
    </div>
    <div
      className="search-title-icon hover-icon"
      onClick={[Function]}
    >
      <i
        className="fas fa-filter fa-2x hover-icon"
      />
    </div>
  </div>
  <form
    className="main-form"
    onSubmit={[Function]}
  >
    <div
      className="exit-filters-icon"
      onClick={[Function]}
    >
      <i
        className="fas fa-times"
      />
    </div>
    <Row
      lg={4}
      noGutters={false}
      sm="4"
      xs={1}
    >
      <Col
        xs="12"
      >
        <Text
          className="form-control input-error"
          errorMsg="Brak tytułu"
          isErrorConditionBlock={[Function]}
          name="q"
          onChange={[Function]}
          placeholder="Tytuł"
        />
      </Col>
      <Col
        xs="12"
      >
        <Text
          className="form-control input-error"
          isErrorConditionBlock={[Function]}
          name="authors"
          onChange={[Function]}
          placeholder="Autorzy"
        />
      </Col>
      <Col
        xs="12"
      >
        <select
          className="browser-default custom-select"
          name="langRestrict"
          onChange={[Function]}
          placeholder="Język"
        >
          <option
            key="0"
            label="Wybierz Język"
            value=""
          />
          <option
            key="1"
            label="Polski"
            value="pl"
          />
          <option
            key="2"
            label="Angielski"
            value="en"
          />
          <option
            key="3"
            label="Włoski"
            value="it"
          />
          <option
            key="4"
            label="Hiszpański"
            value="es"
          />
          <option
            key="5"
            label="Portugalski"
            value="pt"
          />
          <option
            key="6"
            label="Niemiecki"
            value="de"
          />
          <option
            key="7"
            label="Rosyjski"
            value="ru"
          />
          <option
            key="8"
            label="Francuski"
            value="fr"
          />
          <option
            key="9"
            label="Fiński"
            value="fi"
          />
          <option
            key="10"
            label="Szwedzki"
            value="sv"
          />
          <option
            key="11"
            label="Słowacki"
            value="sl"
          />
          <option
            key="12"
            label="Norweski"
            value="no"
          />
        </select>
      </Col>
      <Col
        xs="12"
      >
        <select
          className="browser-default custom-select"
          name="filter"
          onChange={[Function]}
          placeholder="Język"
        >
          <option
            label="Typ e-booka"
            value={null}
          />
          <option
            label="Darmowe"
            value="free-ebooks"
          />
          <option
            label="Płatne"
            value="paid-ebooks"
          />
          <option
            label="Darmowe i płatne"
            value="ebooks"
          />
        </select>
      </Col>
      <Col
        style={
          Object {
            "display": "none",
          }
        }
        xs="12"
      >
        <button
          className="btn btn-success"
          disabled={true}
          onClick={[Function]}
          type="submit"
        >
          Filtruj
        </button>
      </Col>
    </Row>
  </form>
</div>
`;

exports[`<VolumeComponent /> should match snapshot 1`] = `
<div
  className="volume-container"
>
  <div
    className="img-container"
  >
    <img
      alt=""
      src="book.png"
    />
  </div>
  <div
    className="additional-info-container"
  >
    <h2>
      <LinesEllipsis
        basedOn="letters"
        className=""
        component="div"
        ellipsis="..."
        maxLine="2"
        onReflow={[Function]}
        text=""
        trimRight={true}
      />
    </h2>
    <h4
      className="h4-description"
    >
      <LinesEllipsis
        basedOn="letters"
        className=""
        component="div"
        ellipsis="..."
        maxLine="2"
        onReflow={[Function]}
        text="Brak Opisu"
        trimRight={true}
      />
    </h4>
    <h4
      id="autors"
    >
      Autorzy: 
    </h4>
    <h4 />
    <div
      className="price"
    >
      <p
        id="not-for-sale"
      >
        Nie na sprzedaż
      </p>
    </div>
  </div>
</div>
`;
